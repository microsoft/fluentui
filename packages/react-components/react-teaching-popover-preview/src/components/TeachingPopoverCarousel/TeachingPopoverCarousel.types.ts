import * as React from 'react';
import { ComponentProps, ComponentState, EventData, EventHandler, Slot } from '@fluentui/react-utilities';
import { Button } from '@fluentui/react-button';
import { PopoverContextValue } from '@fluentui/react-popover';
import { TeachingPopoverCarouselNavProps } from '../TeachingPopoverCarouselNav/index';

export type TeachingPopoverCarouselSlots = {
  /**
   * The element wrapping carousel pages and navigation.
   */
  root: NonNullable<Slot<'div'>>;

  /**
   * The element wrapping the navigation of the carousel.
   */
  footer: NonNullable<Slot<'div'>>;

  /**
   * The previous button slot.
   */
  previous: Slot<typeof Button>;

  /**
   * The next button slot.
   */
  next: NonNullable<Slot<typeof Button>>;

  /**
   * The page nav slot when using icon navigation.
   */
  nav: Slot<TeachingPopoverCarouselNavProps>;

  /**
   * The page count text slot for paginationType text
   */
  pageCount: Slot<'div'>;
};

export type TeachingPopoverCarouselLayout = 'offset' | 'centered';

// For localization or customization, users may want to modify this for their own purposes
export type TeachingPopoverPageCountChildRenderFunction = (currentPage: number, totalPages: number) => React.ReactNode;
export type TeachingPopoverPageCountRenderType = TeachingPopoverPageCountChildRenderFunction | string;
export type TeachingPopoverStrings = {
  previous: string;
  initialStepText: string;
  next: string;
  finalStepText: string;
  /**
   * Localize the separator between numbers, or use a function to fully override
   */
  pageCountText?: TeachingPopoverPageCountRenderType;
};

export type TeachingPopoverPageChangeData = EventData<'click', React.MouseEvent<HTMLButtonElement>> & {
  /**
   * The page to be set after event has occurred.
   */
  currentPage: number;
};

/**
 * TeachingPopoverCarousel Props
 */
export type TeachingPopoverCarouselProps = Partial<ComponentProps<TeachingPopoverCarouselSlots>> & {
  /**
   * Controls whether buttons will be centered (balanced) or right aligned
   * Defaults to 'centered'.
   */
  layout?: TeachingPopoverCarouselLayout;

  /**
   * Dictates whether pagination uses text or icons
   * Defaults to icons
   */
  paginationType?: 'text' | 'icon';

  /**
   * Strings used to localize carousel functionality
   */
  strings: TeachingPopoverStrings;

  /**
   * Page at which carousel should be initialized to
   */
  defaultCurrentPage?: number;

  /**
   * Callback to notify a page change (can be used to update 'currentPage' externally).
   */
  onPageChange?: EventHandler<TeachingPopoverPageChangeData>;

  /**
   * Callback to notify when the final button step of a carousel has been activated.
   */
  onFinish?: EventHandler<TeachingPopoverPageChangeData>;

  /**
   * Controllable page state
   */
  currentPage?: number;
};

/**
 * TeachingPopoverCarousel State and Context Hooks
 */
export type TeachingPopoverCarouselState = ComponentState<TeachingPopoverCarouselSlots> & {
  totalPages: number;
  setCurrentPage: (page: number) => void;
} & Partial<Pick<PopoverContextValue, 'appearance'>> &
  Pick<TeachingPopoverCarouselProps, 'layout' | 'onPageChange'> &
  Required<Pick<TeachingPopoverCarouselProps, 'currentPage'>>;
