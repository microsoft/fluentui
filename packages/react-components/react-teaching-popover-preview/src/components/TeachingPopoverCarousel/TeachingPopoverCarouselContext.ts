import { createContext, useContextSelector } from '@fluentui/react-context-selector';
import type { ContextSelector, Context } from '@fluentui/react-context-selector';
import { TeachingPopoverCarouselState } from './TeachingPopoverCarousel.types';

const TeachingPopoverCarouselContext: Context<TeachingPopoverCarouselContextValue> = createContext<
  TeachingPopoverCarouselContextValue | undefined
>(undefined) as Context<TeachingPopoverCarouselContextValue>;

const teachingPopoverCarouselContextDefaultValue: TeachingPopoverCarouselContextValue = {
  currentPage: 0,
  setCurrentPage: () => null,
  totalPages: 1, //'One-pager' default state until proven otherwise
  onPageChange: () => null,
};

export const TeachingPopoverCarouselProvider = TeachingPopoverCarouselContext.Provider;

/**
 * Context shared between TeachingPopoverCarousel and its children components
 */
export type TeachingPopoverCarouselContextValues = {
  carousel: TeachingPopoverCarouselContextValue;
};

/**
 * Context shared between TeachingPopoverCarousel and its children components
 */
export type TeachingPopoverCarouselContextValue = Pick<
  TeachingPopoverCarouselState,
  'currentPage' | 'setCurrentPage' | 'totalPages' | 'onPageChange'
>;

export const useTeachingPopoverCarouselContext_unstable = <T>(
  selector: ContextSelector<TeachingPopoverCarouselContextValue, T>,
): T =>
  useContextSelector(TeachingPopoverCarouselContext, (ctx = teachingPopoverCarouselContextDefaultValue) =>
    selector(ctx),
  );

export function useTeachingPopoverCarouselContextValues_unstable(
  state: TeachingPopoverCarouselState,
): TeachingPopoverCarouselContextValues {
  const { currentPage, setCurrentPage, totalPages, onPageChange } = state;

  const carousel = {
    currentPage,
    setCurrentPage,
    totalPages,
    onPageChange,
  };

  return { carousel };
}
