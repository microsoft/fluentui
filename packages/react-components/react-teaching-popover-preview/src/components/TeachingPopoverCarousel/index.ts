export * from './TeachingPopoverCarousel';
export * from './TeachingPopoverCarousel.types';
export * from './renderTeachingPopoverCarousel';
export * from './useTeachingPopoverCarousel';
export * from './useTeachingPopoverCarouselStyles.styles';
export * from './useTeachingPopoverCarouselContextValues';
