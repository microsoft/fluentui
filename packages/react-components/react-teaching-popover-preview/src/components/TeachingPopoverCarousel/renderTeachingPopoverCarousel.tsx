/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { assertSlots } from '@fluentui/react-utilities';
import type {
  TeachingPopoverCarouselState,
  TeachingPopoverCarouselSlots,
  TeachingPopoverCarouselContextValues,
} from './TeachingPopoverCarousel.types';
import { CarouselProvider } from './Carousel/CarouselContext';

/**
 * Render the final JSX of TeachingPopoverCarousel
 */
export const renderTeachingPopoverCarousel_unstable = (
  state: TeachingPopoverCarouselState,
  contextValues: TeachingPopoverCarouselContextValues,
) => {
  assertSlots<TeachingPopoverCarouselSlots>(state);

  return (
    <CarouselProvider value={contextValues.carousel}>
      <state.root />
    </CarouselProvider>
  );
};
