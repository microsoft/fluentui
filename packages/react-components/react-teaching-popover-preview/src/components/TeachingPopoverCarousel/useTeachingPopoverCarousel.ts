import * as React from 'react';
import { getIntrinsicElementProps, slot } from '@fluentui/react-utilities';
import type { TeachingPopoverCarouselProps, TeachingPopoverCarouselState } from './TeachingPopoverCarousel.types';

/**
 * Create the state required to render TeachingPopoverCarousel.
 *
 * The returned state can be modified with hooks such as useTeachingPopoverCarouselStyles_unstable,
 * before being passed to renderTeachingPopoverCarousel_unstable.
 *
 * @param props - props from this instance of TeachingPopoverCarousel
 * @param ref - reference to root HTMLDivElement of TeachingPopoverCarousel
 */
export const useTeachingPopoverCarousel_unstable = (
  props: TeachingPopoverCarouselProps,
  ref: React.Ref<HTMLDivElement>,
): TeachingPopoverCarouselState => {
  return {
    // TODO add appropriate props/defaults
    components: {
      // TODO add each slot's element type or component
      root: 'div',
    },
    // TODO add appropriate slots, for example:
    // mySlot: resolveShorthand(props.mySlot),
    root: slot.always(
      getIntrinsicElementProps('div', {
        ref,
        ...props,
      }),
      { elementType: 'div' },
    ),
  };
};
