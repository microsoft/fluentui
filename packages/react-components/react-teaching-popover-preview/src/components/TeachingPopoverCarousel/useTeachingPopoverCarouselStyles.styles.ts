import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type { TeachingPopoverCarouselSlots, TeachingPopoverCarouselState } from './TeachingPopoverCarousel.types';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { tokens } from '@fluentui/react-theme';

export const teachingPopoverCarouselClassNames: SlotClassNames<TeachingPopoverCarouselSlots> = {
  root: 'fui-TeachingPopoverCarousel',
  footer: 'fui-TeachingPopoverCarousel__footer',
  previous: 'fui-TeachingPopoverCarousel__previous',
  next: 'fui-TeachingPopoverCarousel__next',
  nav: 'fui-TeachingPopoverCarousel__nav',
  pageCount: 'fui-TeachingPopoverCarousel__pageCount',
};

// Todo: Page change animation & styles
const useStyles = makeStyles({
  root: {},
  footer: {
    display: 'flex',
    flexDirection: 'row',
  },
  footerCentered: {
    justifyContent: 'space-between',
    ...shorthands.gap('8px'),
  },
  footerRightAligned: {
    ...shorthands.gap('8px'),
    ':nth-child(0)': {
      marginRight: 'auto',
    },
  },
  previous: {
    minWidth: '96px',
  },
  next: {
    minWidth: '96px',
  },
  brandNext: {
    color: tokens.colorBrandForeground1,
    backgroundColor: tokens.colorNeutralForegroundOnBrand,
    ...shorthands.borderColor(tokens.colorTransparentBackground),
    ':hover': {
      color: tokens.colorCompoundBrandForeground1Hover,
      backgroundColor: tokens.colorNeutralForegroundOnBrand,
    },
    ':hover:active': {
      color: tokens.colorCompoundBrandForeground1Pressed,
      backgroundColor: tokens.colorNeutralForegroundOnBrand,
    },
  },
  brandPrevious: {
    // In brand, this is always 'NeutralForegroundOnBrand'
    color: tokens.colorNeutralForegroundOnBrand,
    backgroundColor: tokens.colorBrandBackground,
    ...shorthands.borderColor(tokens.colorNeutralForegroundOnBrand),
    ':hover': {
      color: tokens.colorNeutralForegroundOnBrand,
      ...shorthands.borderColor(tokens.colorNeutralForegroundOnBrand),
      backgroundColor: tokens.colorBrandBackgroundHover,
    },
    ':hover:active': {
      color: tokens.colorNeutralForegroundOnBrand,
      ...shorthands.borderColor(tokens.colorNeutralForegroundOnBrand),
      backgroundColor: tokens.colorBrandBackgroundPressed,
    },
  },
  nav: {},
  pageCount: { display: 'flex', color: tokens.colorNeutralForeground3, justifyContent: 'center', alignItems: 'center' },
});

/** Applies style classnames to slots */
export const useTeachingPopoverCarouselStyles_unstable = (state: TeachingPopoverCarouselState) => {
  const styles = useStyles();
  const { appearance, layout } = state;

  state.root.className = mergeClasses(teachingPopoverCarouselClassNames.root, styles.root, state.root.className);
  state.footer.className = mergeClasses(
    teachingPopoverCarouselClassNames.footer,
    styles.footer,
    layout === 'centered' ? styles.footerCentered : styles.footerRightAligned,
    state.footer.className,
  );
  if (state.previous) {
    state.previous.className = mergeClasses(
      teachingPopoverCarouselClassNames.previous,
      styles.previous,
      appearance === 'brand' && styles.brandPrevious,
      state.previous.className,
    );
  }
  state.next.className = mergeClasses(
    teachingPopoverCarouselClassNames.next,
    styles.next,
    appearance === 'brand' && styles.brandNext,
    state.next.className,
  );
  if (state.nav) {
    state.nav.className = mergeClasses(teachingPopoverCarouselClassNames.nav, styles.nav, state.nav.className);
  }
  if (state.pageCount) {
    state.pageCount.className = mergeClasses(
      teachingPopoverCarouselClassNames.pageCount,
      styles.pageCount,
      state.pageCount.className,
    );
  }

  return state;
};
