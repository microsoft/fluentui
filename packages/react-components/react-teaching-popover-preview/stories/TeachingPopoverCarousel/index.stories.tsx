import { TeachingPopoverCarousel } from '@fluentui/react-teaching-popover-preview';

import descriptionMd from './TeachingPopoverCarouselDescription.md';
import bestPracticesMd from './TeachingPopoverCarouselBestPractices.md';

export { Default } from './TeachingPopoverCarouselDefault.stories';

export default {
  title: 'Preview Components/TeachingPopoverCarousel',
  component: TeachingPopoverCarousel,
  parameters: {
    docs: {
      description: {
        component: [descriptionMd, bestPracticesMd].join('\n'),
      },
    },
  },
};
