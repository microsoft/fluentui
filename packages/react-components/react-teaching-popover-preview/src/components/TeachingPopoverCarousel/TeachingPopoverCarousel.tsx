import * as React from 'react';
import { useTeachingPopoverCarousel_unstable } from './useTeachingPopoverCarousel';
import { renderTeachingPopoverCarousel_unstable } from './renderTeachingPopoverCarousel';
import { useTeachingPopoverCarouselStyles_unstable } from './useTeachingPopoverCarouselStyles.styles';
import { useCustomStyleHook_unstable } from '@fluentui/react-shared-contexts';
import type { TeachingPopoverCarouselProps } from './TeachingPopoverCarousel.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useTeachingPopoverCarouselContextValues_unstable } from './TeachingPopoverCarouselContext';

/**
 * Define a styled TeachingPopoverCarousel, using the `useTeachingPopoverCarousel_unstable` and `useTeachingPopoverCarouselStyles_unstable`
 * hooks.
 */
export const TeachingPopoverCarousel: ForwardRefComponent<TeachingPopoverCarouselProps> = React.forwardRef(
  (props, ref) => {
    const state = useTeachingPopoverCarousel_unstable(props, ref);

    useTeachingPopoverCarouselStyles_unstable(state);

    useCustomStyleHook_unstable('useTeachingPopoverCarouselStyles_unstable')(state);

    const contextValues = useTeachingPopoverCarouselContextValues_unstable(state);

    return renderTeachingPopoverCarousel_unstable(state, contextValues);
  },
);

TeachingPopoverCarousel.displayName = 'TeachingPopoverCarousel';
