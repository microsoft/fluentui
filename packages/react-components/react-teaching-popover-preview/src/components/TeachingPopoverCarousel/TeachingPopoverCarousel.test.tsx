import * as React from 'react';
import { render } from '@testing-library/react';
import { isConformant } from '../../testing/isConformant';
import { TeachingPopoverCarousel } from './TeachingPopoverCarousel';

describe('TeachingPopoverCarousel', () => {
  isConformant({
    Component: TeachingPopoverCarousel,
    displayName: 'TeachingPopoverCarousel',
    requiredProps: {
      defaultValue: '',
    },
    disabledTests: ['component-has-static-classnames-object'],
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(
      <TeachingPopoverCarousel defaultValue="">Default TeachingPopoverCarousel</TeachingPopoverCarousel>,
    );
    expect(result.container).toMatchSnapshot();
  });
});
