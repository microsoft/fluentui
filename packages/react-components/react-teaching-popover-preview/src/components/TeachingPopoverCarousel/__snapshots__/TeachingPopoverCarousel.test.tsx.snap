// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeachingPopoverCarousel renders a default state 1`] = `
<div>
  <div
    class="fui-TeachingPopoverCarousel"
  >
    Default TeachingPopoverCarousel
    <div
      class="fui-TeachingPopoverCarousel__footer"
    >
      <button
        class="fui-Button fui-TeachingPopoverCarousel__previous"
        type="button"
      >
        Close
      </button>
      <div
        class="fui-TeachingPopoverCarouselNav fui-TeachingPopoverCarousel__nav"
        data-tabster="{\\"groupper\\":{\\"tabbability\\":1},\\"focusable\\":{}}"
        role="list"
        tabindex="0"
      >
        <button
          class="fui-TeachingPopoverCarouselNavButton"
          type="button"
        />
      </div>
      <button
        class="fui-Button fui-TeachingPopoverCarousel__next"
        type="button"
      >
        Finish
      </button>
    </div>
  </div>
</div>
`;
